src/test/scala/scalaz/parsers/ErrorTrackingExampleSpec.scala

```diff
@@ -0,0 +1,286 @@
+
+      override def handleError[A](fa: Eff[A])(f: String => Eff[A]): Eff[A] =
+        EitherT(fa.run.flatMap {
+          case -\/(e) => f(e).run
+          case \/-(_) => fa.run
+        })
+    }
+
+    type WriterState[A] = RWS[Unit, List[Int /\ String], Int, A]
+
+    implicit val writerStateFoldable: Foldable[WriterState] =
+      new Foldable[WriterState] with Foldable.FromFoldr[WriterState] {
+        override def foldRight[A, B](fa: WriterState[A], z: => B)(f: (A, => B) => B): B =
+          scalaz.idInstance.foldr(fa.runZero(())._2, z)(a => b => f(a, b))
+      }
+  }
+
+  trait ParserInstances1 {
+    val monad: Monad[Eff] = implicitly
+  }
+
+  object Example {
+    import Syntax._
+    import TCInstances.naturalTransformationLoop
+
+    import Parser._
+    import Parser.equiv._
+
+    val constantEq: Equiv[Int, Constant] =
+      lift(Constant, _.value)
+
+    val constantExpressionEq: Equiv[Constant, Expression] =
+      liftPartial("Expected: Constant")(
+        { case a               => a },
+        { case n @ Constant(_) => n }
+      )
+
+    def operationExpressionEq(
+      op: Operator
+    ): Equiv[Expression /\ (Operator /\ Expression), Expression] =
+      liftPartial("Expected: Operation")(
+        { case (e1, (`op`, e2))        => Operation(e1, op, e2) },
+        { case Operation(e1, `op`, e2) => e1 -> (op -> e2) }
+      )
+
+    val digit: Codec[Char] = char ∘ ensure("Expected: [0-9]")(_.isDigit)
+
+    val plus: Codec[Operator] = char ∘ liftPartial("Expected: '+'")(
+      { case '+' => Add },
+      { case Add => '+' }
+    )
+
+    val star: Codec[Operator] = char ∘ liftPartial("Expected: '*'")(
+      { case '*' => Mul },
+      { case Mul => '*' }
+    )
+
+    // todo: how to get this error?
+    val integer: Codec[Int] = digit.many1("Expected at least one digit") ∘ lift(
+      chars => new String(chars.toArray).toInt,
+      int => int.toString.toList
+    )
+
+    val constant: Codec[Constant] = integer ∘ constantEq
+
+    val case0: Codec[Expression] = constant ∘ constantExpressionEq
+
+    // todo: how to get this error?
+    val case1: Codec[Expression] = (case0 ~ (star ~ case0).many) ∘ foldl("hm...")(
+      operationExpressionEq(Mul)
+    )
+
+    val case2: Codec[Expression] = (case1 ~ (plus ~ case1).many) ∘ foldl("hmm..")(
+      operationExpressionEq(Add)
+    )
+
+    lazy val expression: Codec[Expression] = case2
+  }
+
+  def parse(s: String): (List[Int /\ String], String \/ (String, Syntax.Expression)) = {
+    val (log, res, _) = Example.expression
+      .parse(s.toCharArray.toList.zipWithIndex)
+      .run
+      .map(_.toEither.map { case (cc, exp) => cc.unzip._1.mkString -> exp })
+      .runZero(())
+    log -> res
+  }
+
+  def parsingResult(s: String): String \/ (String, Syntax.Expression) =
+    parse(s)._2
+
+  def lastParsingError(s: String): Option[Int /\ String] =
+    Option(parse(s)._1.distinct.groupBy(_._1).maxBy(_._1))
```

Comment:
kinda minor but the functions on 141, 142 and 145 are a bit messy, I wonder if these are common enough to justify to putting them some where in the parsers core for sharing.

Reply:
Yes, I agree they are messy.
These functions extract some data from the parser output, so it is easy to read test code below. But since parser output depends on the effect being used, and the effect is specific for this example, we cannot reuse this code or move it outside.